File: client/src/messages/background.ts
import type { BackgroundLayer } from "../models/background";

/** Request to add a background layer */
export interface BackgroundAddLayerRequestMessage {
  type: "request_background_add_layer";
  layer: BackgroundLayer;
}

/** Add a background layer */
export interface BackgroundAddLayerResponseMessage {
  type: "background_add_layer";
  layer: BackgroundLayer;
}

/** Request to select a background layer */
export interface BackgroundSelectLayerRequestMessage {
  type: "request_background_select_layer";
  id: string;
}

/** Select a background layer */
export interface BackgroundSelectLayerResponseMessage {
  type: "background_select_layer";
  id: string;
}

/** Request to rename a background layer */
export interface BackgroundRenameLayerRequestMessage {
  type: "request_background_rename_layer";
  id: string;
  name: string;
}

/** Rename a background layer */
export interface BackgroundRenameLayerResponseMessage {
  type: "background_select_layer";
  id: string;
  name: string;
}

/** Request to delete a background layer */
export interface BackgroundDeleteLayerRequestMessage {
  type: "request_background_delete_layer";
  id: string;
}

/** Select a background layer */
export interface BackgroundDeleteLayerResponseMessage {
  type: "background_delete_layer";
  id: string;
}

export type BackgroundRequestMessage =
  | BackgroundAddLayerRequestMessage
  | BackgroundSelectLayerRequestMessage
  | BackgroundRenameLayerRequestMessage
  | BackgroundDeleteLayerRequestMessage;

export type BackgroundResponseMessage =
  | BackgroundAddLayerResponseMessage
  | BackgroundSelectLayerResponseMessage
  | BackgroundRenameLayerResponseMessage
  | BackgroundDeleteLayerResponseMessage;

File: server/Source/Messages/Background.cs
using Demiplane.Model;
using Newtonsoft.Json;

namespace Demiplane.Messages;

public class BackgroundAddLayerRequestMessage(BackgroundLayer layer) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "request_background_add_layer";

    [JsonProperty(Required = Required.Always)]
    public BackgroundLayer layer = layer;
}

public class BackgroundAddLayerResponseMessage(BackgroundLayer layer) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "background_add_layer";

    [JsonProperty(Required = Required.Always)]
    public BackgroundLayer layer = layer;
}

public class BackgroundSelectLayerRequestMessage(string id) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "request_background_select_layer";

    [JsonProperty(Required = Required.Always)]
    public string id = id;

}

public class BackgroundSelectLayerResponseMessage(string id) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "background_select_layer";

    [JsonProperty(Required = Required.Always)]
    public string id = id;
}

public class BackgroundRenameLayerRequestMessage(string id, string name) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "request_background_rename_layer";

    [JsonProperty(Required = Required.Always)]
    public string id = id;

    [JsonProperty(Required = Required.Always)]
    public string name = name;
}

public class BackgroundRenameLayerResponseMessage(string id, string name) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "background_rename_layer";

    [JsonProperty(Required = Required.Always)]
    public string id = id;

    [JsonProperty(Required = Required.Always)]
    public string name = name;
}


public class BackgroundDeleteLayerRequestMessage(string id) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "request_background_delete_layer";

    [JsonProperty(Required = Required.Always)]
    public string id = id;

}

public class BackgroundDeleteLayerResponseMessage(string id) : Message
{
    [JsonProperty(Required = Required.Always)]
    public string type = "background_delete_layer";

    [JsonProperty(Required = Required.Always)]
    public string id = id;
}
